lib/oga/xml/entities.rb

```diff
@@ -90,7 +90,7 @@ def self.decode(input, mapping = DECODE_MAPPING)
 
         if input.include?(AMPERSAND)
           input = input.gsub(CODEPOINT_ENTITY) do |match|
-            [$1 ? Integer($2, 16) : Integer($2, 10)].pack('U*')
+            [$1.start_with?('x') ? Integer($1[1..-1], 16) : Integer($1, 10)].pack('U*')
```

Comment:
So the reason for the `$1 ? ... : ...` bit of code is that this is basically a hack to see if the entity starts with `x` or not (hence it's in a match group). While this will still allocate a String for the "x" it removes the need for calling `start_with?` as well as the need for the string slicing.

Honestly it might be better to break the regex and the corresponding find/replace up in two blocks. The first one would replace regular numeric entities (since those are probably more common) with the second block replacing the hexadecimal entities. Something like this:

```ruby
if input.include?(AMPERSAND)
  input = input.gsub(NUMERIC_CODE_POINT_ENTITY) do |match|
    ...
  end
end

# This prevents a call to gsub when there are no entities left
if input.include?(AMPERSAND)
  input = input.gsub(HEX_CODE_POINT_ENTITY) do |match|
    ...
  end
end
```

This would remove the need for `start_with?` and the string slicing as well as the ternary itself.